mariadb 10.2: support virtual column

naoa opened the pull request:
#159 

supported virtual(computed) column of MariaDB.

Virtual(computed) columns starting with 5.2, 10.2.1
https://mariadb.com/kb/en/mariadb/virtual-computed-columns/

JSON functions starting with 10.2.3
https://mariadb.com/kb/en/mariadb/json-functions/

added JSON as an alias for TEXT with 10.2.7
https://jira.mariadb.org/browse/MDEV-9144

kou commented on mysql-test/mroonga/include/mroonga/skip_mariadb_10_1_or_earlier.inc:
```diff
@@ -0,0 +1,24 @@
+--source ../../include/mroonga/check_version.inc
+--source ../../include/mroonga/check_mariadb.inc
+
+if ($version_10_1_or_earlier) {
```
Can we use `!$version_10_2_or_later`?

naoa commented:
> Can we use !$version_10_2_or_later?

It could be used with MariaDB 10.2.7. I've fixed it and squashed.

Btw, currently, Mroonga creates persistent column even if created virtual columns with ``VIRTUAL`` flag.
Maybe, we should avoid creation with this flag as not supported.
(I'm not sure but MySQL 5.7 don't create persistent column with ``VIRTUAL``. I'll also check it after change MySQL 5.7.19 from MariaDB 10.2.7.)

```SQL
...
message VARCHAR(255) GENERATED ALWAYS AS (json_extract(`record`, '$.message')) VIRTUAL,
...
```

```bash
dump
...
column_create logs message COLUMN_SCALAR ShortText
...
```

```sql
MariaDB [repair_test]> DROP TABLE IF EXISTS logs;
Query OK, 0 rows affected (0.01 sec)

MariaDB [repair_test]> CREATE TABLE logs (
    -> id INT,
    -> record JSON,
    -> message VARCHAR(255) GENERATED ALWAYS AS (json_extract(`record`, '$.message')) VIRTUAL,
    -> FULLTEXT INDEX(message) comment 'tokenizer "TokenBigramSplitSymbolAlphaDigit"'
    -> ) ENGINE=Mroonga DEFAULT CHARSET=utf8mb4;
Query OK, 0 rows affected, 1 warning (0.00 sec)

MariaDB [repair_test]> SELECT mroonga_command("dump");
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| mroonga_command("dump")                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| table_create logs TABLE_NO_KEY
column_create logs id COLUMN_SCALAR Int32
column_create logs message COLUMN_SCALAR ShortText
column_create logs record COLUMN_SCALAR LongText

table_create logs#message TABLE_PAT_KEY ShortText --default_tokenizer TokenBigramSplitSymbolAlphaDigit --normalizer NormalizerAuto

table_create mroonga_operations TABLE_NO_KEY
column_create mroonga_operations record COLUMN_SCALAR UInt32
column_create mroonga_operations table COLUMN_SCALAR ShortText
column_create mroonga_operations type COLUMN_SCALAR ShortText

column_create logs#message index COLUMN_INDEX|WITH_POSITION logs message |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

MariaDB [repair_test]> INSERT INTO logs(id, record) VALUES (1, '{"level": "info", "message": "start"}');
Query OK, 1 row affected (0.00 sec)

MariaDB [repair_test]> INSERT INTO logs(id, record) VALUES (2, '{"level": "info", "message": "restart"}');
Query OK, 1 row affected (0.00 sec)

MariaDB [repair_test]> INSERT INTO logs(id, record) VALUES (3, '{"level": "warn", "message": "abort"}');
Query OK, 1 row affected (0.00 sec)

MariaDB [repair_test]> SELECT * FROM logs WHERE MATCH(message) AGAINST("ar" IN BOOLEAN MODE);
+------+-----------------------------------------+-----------+
| id   | record                                  | message   |
+------+-----------------------------------------+-----------+
|    1 | {"level": "info", "message": "start"}   | "start"   |
|    2 | {"level": "info", "message": "restart"} | "restart" |
+------+-----------------------------------------+-----------+
2 rows in set (0.03 sec)
```

naoa commented:
I've implemented to support ``VIRTUAL`` type of MariaDB without index.

If this PR is merged, I'll create PR for it.

https://github.com/mroonga/mroonga/commit/6e5fb85f09b5c81b10c5c243082851340cb4545b
https://github.com/mroonga/mroonga/commit/c4ddbcbab846d8c2fb50e970fbc9d56a53b10ca7

kou commented:
Thanks!
Please create a new pull request!

kou commented:
It seems that we need more works to support inplace `ADD COLUMN GENERATED`.
I've changed tests and disabled inplace `ADD COLUMN GENERATED` for now to get the correct result: d40e5d7acd5bdbc66c12d63a01585867e3efbecc

kou commented:
Inserted data before inplace `ADD COLUMN GENERATED` are missing.

naoa commented:
Oh, i see.
If I don't have use case for using generated column and it seems to be difficult, it may not do it for a while.

kou commented:
One of use cases:

  * An user stores log data as JSON. They are just stored not searched.
  * The user want to start searching the stored data.

I don't know how difficult to implement it because I don't look into it...
